### DISCUSSION TITLE
How to speed up my Tor network simulation with custom processes?

### DISCUSSION THREAD

**Discussion by higher-order** [shadow/shadow#3415] [Q&A] [answered]
I'm using tornettools to generate Tor networks of 2%, 5%, and 10% scales, and hoping to simulate them for 3 hours. I added processes to some Tor clients to run Python programs that I wrote. These programs use the `stem` library to listen to Tor control port events, compute the throughput with different relays, and report them via HTTP requests to the bandwidth authority (bwauth). bwauth is a host in the simulation that runs a Python Flask web server that handles these reports, and tries to write sensible values to the bandwidth file at `shadow.data/hosts/bwauthority/v3bw`. Each client starts reporting at different time, but all report every minute. bwauth writes the v3bw every minute. (I assume if I write `sleep(60)` in the code, it will sleep 60 seconds in simulation time, and not real time, right?)

For a 5% network and a 3-hour simulation, the original network took about 24 hours, and the experimental network (the one with clients reporting to bwauth) took about 130 hours. Surprisingly, the original network consumed 250+ GiB of RAM, while the experimental network consumed around 110 GiB, even though the experimental network is an original network with additional processes. I'm attaching the plot here. "orig2crowd" is the experimental network. "dynamic random (failed)" is an original network with an additional process that keeps writing random values to the v3bw every minute, and terminated at the 2.5 hours mark due to RAM exhaustion.

[tornet.plot.pages.pdf](https://github.com/user-attachments/files/17051081/tornet.plot.pages.pdf)

I'd like to speed up the experimental simulation. I'm not allowed to disable side-channel mitigation. So I first looked at Shadow's config options that I can tune [here](https://shadow.github.io/docs/guide/perf_config_options.html). I didn't try most of the options since they are described as typically harmful and only beneficial at certain circumstances. Among the ones I adjusted:
* I increased `heartbeat_interval` and `host_heartbeat_interval` from 1 sec to 5 sec, since I aim for a 3-hour run. And since I assume the heartbeats will also be written to files on the drive like everything else, making it less frequent would reduce the amount of I/O. But the simulations still took a lot of time, as shown in the plot. And I don't know if increasing these intervals have any negative effects?
* Will it make sense to increase `runahead`? The docs says "This option effectively sets a minimum network latency. Increasing this value will allow for better simulation parallelisation and possibly better run time performance." The default is 1 ms. I assume for a Tor network, the minimum network latency will be much greater than 1 ms, since Tor is slow? Or am I misunderstanding what it is?

I then hoped to optimize my programs so that they simulate faster.
* The clients use stem to get the `CIRC_BW` event from the Tor control port. When I'm testing outside of Shadow, each client prints many such events per second, scrolling the screen pretty rapidly. The printing is commented out in the simulation to prevent too much I/O. I wonder if too many control port events like this is a bottleneck for simulation run time?
* The bwauth runs a Python Flask server to handle the reports from clients.
  * Is Flask a sub-optimal server for Shadow?
  * I also made Flask single-threaded with `app.run(threaded=False, ...)`. Is that likely to hurt the performance much?
  * But if I make it multi-threaded, it'll probably need to change the current dictionary database to an ACID database like PostgreSQL. Will that create too much complication for Shadow?

I'm also wondering if a slow network will be slower to simulate? If bwauth is not writing good values to the v3bw, the Tor network will probably be less performant overall. Will that cause the simulation to also be slower?

Finally, I tried to follow the [profiling section](https://shadow.github.io/docs/guide/profiling.html) in the docs. I ran `perf record -p <PID>`  to attach to the Shadow process after the clients and bwauth all start. I then ran `perf report` and dumped it into a text file. But I can't make much sense out of it. The top few Shadow symbols seem to be related to waiting, like `std::thread::yield_now`, `LatchWaiter::wait`, and so on. Many other symbols are shown as addresses in the kernel, and I don't know if it will help to figure out what they are, and how to do so. I'm attaching the dumped `perf report` here.

[perf-report-orig2crowd_net_0.02_3600s.txt](https://github.com/user-attachments/files/17051287/perf-report-orig2crowd_net_0.02_3600s.txt)

I'd love to hear your insights on this, thank you.

**Comment by sporksmith** [comment]
> (I assume if I write sleep(60) in the code, it will sleep 60 seconds in simulation time, and not real time, right?)

Correct

> Surprisingly, the original network consumed 250+ GiB of RAM, while the experimental network consumed around 110 GiB, even though the experimental network is an original network with additional processes.

That is surprising...

> I increased heartbeat_interval and host_heartbeat_interval from 1 sec to 5 sec, since I aim for a 3-hour run. And since I assume the heartbeats will also be written to files on the drive like everything else, making it less frequent would reduce the amount of I/O. But the simulations still took a lot of time, as shown in the plot. And I don't know if increasing these intervals have any negative effects?

There shouldn't be any negative effects other than having less frequent data points in shadow's output. I think tornettools uses  these for some of its graphs, but giving the length of time that you're simulating, it probably makes sense to set these relatively high. e.g. you could try 1m.

> Will it make sense to increase runahead? The docs says "This option effectively sets a minimum network latency. Increasing this value will allow for better simulation parallelisation and possibly better run time performance." The default is 1 ms. I assume for a Tor network, the minimum network latency will be much greater than 1 ms, since Tor is slow? Or am I misunderstanding what it is?

This is talking about the latency between individual hosts. While an end to end Tor connection will typically have a latency much higher than 1 ms, some relays may have fairly low latency links to eachother. You can look at your network graph to get an idea of which links increasing this setting may affect and by how much.

> The clients use stem to get the CIRC_BW event from the Tor control port. When I'm testing outside of Shadow, each client prints many such events per second, scrolling the screen pretty rapidly. The printing is commented out in the simulation to prevent too much I/O. I wonder if too many control port events like this is a bottleneck for simulation run time?

It's possible. See below.

> The bwauth runs a Python Flask server to handle the reports from clients.
>    Is Flask a sub-optimal server for Shadow?
>   I also made Flask single-threaded with app.run(threaded=False, ...). Is that likely to hurt the performance much?
>    But if I make it multi-threaded, it'll probably need to change the current dictionary database to an ACID database like PostgreSQL. Will that create too much complication for Shadow?

I'm not very familiar with Flask, but running it single-threaded is probably the right approach. Shadow emulates a single-core CPU on each host, so increasing the number of threads is unlikely to help performance.

>  I wonder if too many control port events like this is a bottleneck for simulation run time?

It's possible; see below.

> I'm also wondering if a slow network will be slower to simulate? If bwauth is not writing good values to the v3bw, the Tor network will probably be less performant overall. Will that cause the simulation to also be slower?

It's possible. e.g. if there are more errors, timeouts, retries etc then there may be more work being done in the simulation overall.

> Finally, I tried to follow the [profiling section](https://shadow.github.io/docs/guide/profiling.html) in the docs. I ran perf record -p <PID> to attach to the Shadow process after the clients and bwauth all start.

Using the `-p` option means perf will record *only* events in that process. Based on the results you shared, it doesn't look like most of the CPU time is being spent in shadow itself; it's probably being spent inside some processes inside the simulation.

You could try running `perf` without `-p` to record system-wide, though it might be easier to first use a tool like `top` or `htop` to first identify which processes are using the most CPU.

**Reply by higher-order** [reply]
These are all very helpful, thanks! Just to compare if anything is optimized, I plan to run another experimental simulation, but set the two heartbeat intervals to 60 sec. Assuming the `atlas_v201801.shadow_v2.gml` file is the network graph, I'll then write a script to find the lowest latency among its edges, divide it by 1000 to convert μs to ms, and set `runahead` to slightly lower than that. That will improve the performance without affecting the network characteristics, right?

As it runs, I'll also use `htop` to identify the expensive processes, and then use `perf record -p <PID>` on those processes. Since the run time will be long, I'll probably also have the chance to do `perf record` system-wide too.

**Reply by sporksmith** [reply]
> Assuming the atlas_v201801.shadow_v2.gml file is the network graph, I'll then write a script to find the lowest latency among its edges, divide it by 1000 to convert μs to ms, and set runahead to slightly lower than that. That will improve the performance without affecting the network characteristics, right?

cc @robgjansen @stevenengler 

One other possible caveat - I think this latency *might* also affect "self" connections, including e.g. a socks connections and control port connections between tor and local applications. Rob, Steven - do you happen to know whether this is true? I think theoretically it shouldn't *have* to since the receiver is the same host, but I think I recall that it might not be easy to avoid in our current network stack...

> As it runs, I'll also use htop to identify the expensive processes, and then use perf record -p <PID> on those processes. Since the run time will be long, I'll probably also have the chance to do perf record system-wide too.

Sounds good. You'll probably also want to record stacks in the samples for the results to be very useful (`--call-graph`). The data files grow pretty quickly, especially with stack captures enabled, so you'll probably just want to record a few seconds at a time.

**Reply by robgjansen** [reply]
> I think this latency might also affect "self" connections, including e.g. a socks connections and control port connections between tor and local applications. Rob, Steven - do you happen to know whether this is true?

No, I think if it detects that the destination is the same host as the source, it does not even go through the routing layer and will not have the normal latency applied to it. In other words, those "local" events get placed in the local event queue and are processed during the current run interval, rather than being placed in the "remote" event queue and subject to runahead limits.

**Reply by robgjansen** [reply]
Note that changing the runahead only has the _potential_ to improve performance, but actually I don't expect that it will have much of an effect in our Tor simulations. The runahead makes it so that there is more events that can possibly be parallelized. But if the default runahead already produces a fairly equal load distribution across the CPU cores (which tends to be the case in the Tor simulations in my experience), then increasing the runahead will have minimal effect.

It all depends on the density of the events in a run interval. Our usual Tor simulations have enough work across enough hosts that each CPU core tends to be at full usage already even with the default runahead. You can see this in htop during the main parts of the simulation (after bootstrapping), where typically all of the assigned CPUs run at 100%. That being said, if you are adding a small number of hosts that cause unequal work in some intervals, such that some of the CPUs go through periods of low utilization, then increasing the runahead could help a bit.

**Reply by sporksmith** [reply]
> I'll then write a script to find the lowest latency among its edges, divide it by 1000 to convert μs to ms, and set runahead to slightly lower than that.

Actually I think shadow itself effectively does this already. The `--runahead` command line option is to force some higher value than would be calculated this way. See https://github.com/shadow/shadow/blob/86c19651c9471dff52a944787f74bc0aa754bd79/src/main/core/runahead.rs#L48-L56

**Reply by stevenengler** [reply]
> > I'll then write a script to find the lowest latency among its edges, divide it by 1000 to convert μs to ms, and set runahead to slightly lower than that.
> 
> Actually I think shadow itself effectively does this already. The `--runahead` command line option is to force some higher value than would be calculated this way. See https://github.com/shadow/shadow/blob/86c19651c9471dff52a944787f74bc0aa754bd79/src/main/core/runahead.rs#L48-L56

Also Shadow only considers the graph nodes that actually have hosts attached to them so that you're not penalized by low latencies on edges that are never used. So Shadow's runahead calculation only considers a subset of the graph.

**Reply by higher-order** [reply]
Those are very informative, thanks all. Since it seems like I should just leave `runahead` as it is, I'll run a smaller simulation, probably a 2% network for 1 hour, with 60 sec (host) heartbeat intervals and focus on using `perf` to see if some parts of my custom processes are taking a lot of time in the simulation. I'll probably learn how to use perf effectively first.

**Reply by stevenengler** [reply]
> > I think this latency might also affect "self" connections, including e.g. a socks connections and control port connections between tor and local applications. Rob, Steven - do you happen to know whether this is true?
>
> No, I think if it detects that the destination is the same host as the source, it does not even go through the routing layer and will not have the normal latency applied to it. In other words, those "local" events get placed in the local event queue and are processed during the current run interval, rather than being placed in the "remote" event queue and subject to runahead limits.

Technically iirc there's only a [single event queue](https://github.com/shadow/shadow/blob/86c19651c9471dff52a944787f74bc0aa754bd79/src/main/host/host.rs#L731-L750) now, but the behaviour should be the same. New local events should be added to the event queue at t=now(), which puts it at the start of the queue. The host just keeps grabbing the next event until the next event is past the runahead limit (or whatever the manager decides is the limit), and then it stops.

There is bug #2677 where the host's bandwidth limits can theoretically prevent localhost packets from arriving on time. But this is unrelated to latency.

**Comment by robgjansen** [comment]
>> Surprisingly, the original network consumed 250+ GiB of RAM, while the experimental network consumed around 110 GiB, even though the experimental network is an original network with additional processes.

> That is surprising...

Based on the graphs, it seems to me that everything was running as expected until you started modifying the v3bw file which I guess was at the 30 minute mark into the simulation. At that point, you probably introduced sub-optimal load balancing weights, which caused more of the downloads to fail, and slowed down the other ones that succeeded.

Your experimental versions for sure would be doing extra work---all of those extra client tasks reading tor controller output, processing it, etc. will require extra CPU time and more RAM---but it will also reduce some work if many of the downloads are stalling and not transferring the same number of bytes in total compared to the original. The "Transfer Time" and "Relay Goodput" CDFs do show that the orig2crowd version is transferring slightly fewer bytes in total, but not a crazy high amount. So I would not expect that this would be the primary cause of the increase in RAM and CPU runtime. Thus, my next best guess would be to point at your new python processes.

One thing that puzzles me is that the first two graphs (1. RAM vs Real Time, and 2. RAM vs Simulation Time) agree in how much total RAM is being used by original, but do not agree in how much RAM is being used by your orig2crowd and dynamic random variants. Graph 1. says your variants are using much more RAM than graph 2. What is going on here, and which one is correct?

Here is where the memory info is parsed:
https://github.com/shadow/tornettools/blob/9716a8682483f363e4bc9f9503f6871536e7b846/tornettools/parse_rusage.py#L126

I wonder if it took the `else` block here, which you could check in the logs:

https://github.com/shadow/tornettools/blob/9716a8682483f363e4bc9f9503f6871536e7b846/tornettools/parse_rusage.py#L130-L136

**Reply by higher-order** [reply]
Your guess is exactly right. The bwauth process starts running and starts handling reports at the 1830th second, which seems to be align with where the plotted line gets steeper. And good catch on the RAM mismatch in the plot. So I guess orig2crowd is indeed using more RAM than original, though not by too much, fortunately. I don't know what causes the mismatch though. If it took the else branch in the code, it would've logged a warning in the log files, right? But searching for "WARN" in the parsing and plotting logs with vim showed no match.

**Reply by higher-order** [reply]
Regarding the impact of a slower network on the simulation, is my understanding correct that, if the modified v3bw causes some downloads to fail, it may reduce some work to be simulated, and thus reduce run time and RAM usage? But if the modified v3bw just causes some downloads to be slower, it will increase the run time? I read but didn't fully understand the Shadow papers. But I thought I read something about how Shadow will simulate the network latency by waiting. So if the network is slower, Shadow will wait more, causing the run time to increase?

If that is true then maybe there are two reasons behind the siginificantly increased run time (from ~24 hours to ~130 hours): A slower network, and costly custom processes? Of course, I may have completely misunderstood Shadow.

**Reply by robgjansen** [reply]
> if the modified v3bw causes some downloads to fail, it may reduce some work to be simulated, and thus reduce run time and RAM usage? But if the modified v3bw just causes some downloads to be slower, it will increase the run time?

That is my guess, yes. It all depends on the download timeout values. A download timeout will essentially "cancel" the remainder of the download, meaning that extra work of transferring those bytes is canceled.

But again, I do not expect we would see 5x increase in runtime based on the graphs you shared. From the relay goodput graphs, it looks like a similar amount of bytes are being sent through all the relays. This is why I suspect something else to be the main primary cause, and perhaps download differences would cause a minor secondary effect.

**Reply by higher-order** [reply]
Understood, let me profile the custom processes in the next few simulations.

**Comment by higher-order** [comment, marked as answer]
Solved [here](https://github.com/shadow/shadow/discussions/3481#discussioncomment-12013384).
